lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php

```diff
@@ -0,0 +1,832 @@
+
+		$this->groupsAddableByGroup['helper'] = array('rollback', 'bot', 'sysop', 'bureaucrat', 'chatmoderator', 'threadmoderator');
+		$this->groupsRemovableByGroup['helper'] = array('rollback', 'bot', 'sysop', 'bureaucrat', 'chatmoderator', 'threadmoderator');
+
+		$this->groupsAddableByGroup['sysop'] = array('chatmoderator', 'threadmoderator');
+		$this->groupsRemovableByGroup['sysop'] = array('chatmoderator', 'threadmoderator');
+		$this->groupsSelfRemovableByGroup['sysop'] = array('sysop');
+
+		$this->groupsAddableByGroup['content-reviewer'] = array('content-reviewer');
+		$this->groupsRemovableByGroup['content-reviewer'] = array('content-reviewer');
+
+		$this->groupsAddableByGroup['vstf'] = array('rollback', 'bot');
+		$this->groupsRemovableByGroup['vstf'] = array('rollback', 'bot');
+		$this->groupsSelfAddableByGroup['vstf'] = array('sysop');
+		$this->groupsSelfRemovableByGroup['vstf'] = array('sysop', 'bureaucrat');
+
+		//the $wgXXXLocal variables are loaded from wiki factory - we should use it as is
+		if ( !empty( $wgAddGroupsLocal ) )
+			$this->groupsAddableByGroup = array_merge( $this->groupsAddableByGroup, $wgAddGroupsLocal );
+		if ( !empty( $wgRemoveGroupsLocal ) )
+			$this->groupsRemovableByGroup = array_merge( $this->groupsRemovableByGroup, $wgRemoveGroupsLocal );
+		if ( !empty( $wgGroupsAddToSelfLocal ) )
+			$this->groupsSelfAddableByGroup = array_merge( $this->groupsSelfAddableByGroup, $wgGroupsAddToSelfLocal );
+		if ( !empty( $wgGroupsRemoveFromSelfLocal ) )
+			$this->groupsSelfRemovableByGroup = array_merge( $this->groupsSelfRemovableByGroup, $wgGroupsRemoveFromSelfLocal );
+
+		$this->groupsAddableByGroup['util'] = array_diff( $this->getExplicitGroups(),
+			array_merge( [ 'wikifactory', 'content-reviewer' ], $this->getImplicitGroups() ) );
+		$this->groupsRemovableByGroup['util'] = array_diff( $this->getExplicitGroups(), $this->getImplicitGroups() );
+
+		global $wgDevelEnvironment;
+		if ( !empty( $wgDevelEnvironment )) {
+			$this->groupsAddableByGroup['staff'] = $this->getExplicitGroups();
+			$this->groupsRemovableByGroup['staff'] = $this->getExplicitGroups();
+			$this->groupsSelfAddableByGroup['staff'] = $this->getExplicitGroups();
+			$this->groupsSelfRemovableByGroup['staff'] = $this->getExplicitGroups();
+		}
+	}
+
+	public function getGlobalGroups() {
+		return $this->globalGroups;
+	}
+
+	public function getImplicitGroups() {
+		return $this->implicitGroups;
+	}
+
+	public function getExplicitGroups() {
+		return $this->explicitGroups;
+	}
+
+	public function getExplicitUserGroups( $userId ) {
+		if ( empty( $userId ) ) {
+			return [];
+		}
+
+		return array_unique( array_merge (
+			$this->getExplicitLocalUserGroups( $userId ),
+			$this->getExplicitGlobalUserGroups( $userId )
+		) );
+	}
+
+	public function getExplicitLocalUserGroups( $userId ) {
+		$this->loadLocalGroups( $userId );
+		return $this->getLocalUserGroupsArray( $userId );
```

Comment:
I'd suggest naming those methods in such a way it's more obvious which ones load the data and which one don't
I wouldn't guess which on of them does: `getExplicitLocalUserGroups` or `getLocalUserGroupsArray`